vllm/model_executor/layers/sampler.py

```diff
@@ -268,161 +293,230 @@ def _apply_top_p_top_k(
 def _get_topk_logprobs(
```

Comment:
I would also suggest copying over the vectorized `_apply_penalties` code from my PR, as is noticeably faster than the current numpy-based solution.